src/Build/Collections/PropertyDictionary.cs

```diff
@@ -423,13 +497,19 @@ public T GetProperty(string name, int startIndex, int endIndex)
         /// </summary>
         IEnumerator<KeyValuePair<string, T>> IEnumerable<KeyValuePair<string, T>>.GetEnumerator()
         {
-            lock (_properties)
+            _lock.EnterReadLock();
```

Comment:
This method was and continues to be highly suspect.  Maybe it's fine because you control all the callers as well, but the lock is going to be held while this method yields.  In other words, when I do:
```C#
PropertyDictionary pd = ...;
var e = pd.GetEnumerator();
e.MoveNext();
... // lock is held here!
```